pallets/payment-streams/src/utils.rs

```diff
@@ -0,0 +1,490 @@
+use frame_support::ensure;
+use frame_support::pallet_prelude::DispatchResult;
+use frame_support::sp_runtime::{
+    traits::{CheckedMul, CheckedSub, Zero},
+    ArithmeticError, DispatchError,
+};
+use frame_support::traits::{
+    fungible::{Inspect, InspectHold, Mutate, MutateHold},
+    tokens::{Fortitude, Precision, Preservation},
+    Get,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use sp_runtime::traits::Convert;
+
+use crate::*;
+use storage_hub_traits::PaymentStreamsInterface;
+
+macro_rules! expect_or_err {
+    // Handle Option type
+    ($optional:expr, $error_msg:expr, $error_type:path) => {{
+        match $optional {
+            Some(value) => value,
+            None => {
+                #[cfg(test)]
+                unreachable!($error_msg);
+
+                #[allow(unreachable_code)]
+                {
+                    Err($error_type)?
+                }
+            }
+        }
+    }};
+    // Handle boolean type
+    ($condition:expr, $error_msg:expr, $error_type:path, bool) => {{
+        if !$condition {
+            #[cfg(test)]
+            unreachable!($error_msg);
+
+            #[allow(unreachable_code)]
+            {
+                Err($error_type)?
+            }
+        }
+    }};
+}
+
+impl<T> Pallet<T>
+where
+    T: pallet::Config,
+{
+    /// This function holds the logic that checks if a payment stream can be created and, if so, stores the payment stream in the PaymentStreams mapping
+    /// and holds the necessary balance from the sender if it's its first payment stream
+    ///
+    /// Note: Maybe we should add a check to make sure the user has enough balance to pay for at least X amount of blocks?
+    pub fn do_create_payment_stream(
+        bsp_account: &T::AccountId,
+        user_account: &T::AccountId,
+        rate: BalanceOf<T>,
+    ) -> DispatchResult {
+        // Get the BSP ID of the BSP account
+        let bsp_id = <T::Providers as storage_hub_traits::ProvidersInterface>::get_provider(
+            bsp_account.clone(),
+        )
+        .ok_or(Error::<T>::NotABackupStorageProvider)?;
+
+        // Check that a payment stream between that BSP and user does not exist yet
+        ensure!(
+            !PaymentStreams::<T>::contains_key(bsp_id, user_account),
+            Error::<T>::PaymentStreamAlreadyExists
+        );
+
+        // Check that the user is not flagged as without funds
+        ensure!(
+            !UsersWithoutFunds::<T>::contains_key(user_account),
+            Error::<T>::UserWithoutFunds
+        );
+
+        // Check if the user is already registered and, if not, try to hold the deposit
+        let mut user_payment_streams_count = RegisteredUsers::<T>::get(user_account);
+        if user_payment_streams_count == 0 {
```

Comment:
Shouldn't we hold an amount based on the number of payment streams the user has?

Reply:
Even better, we hold it separately for each stream according to the details of each one now!